### PR TITLE
pointer_offset_size functions now use optional

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/c_typecheck_expr.cpp [hunk 1]
```diff
@@ -695,11 +695,18 @@ void c_typecheck_baset::typecheck_expr_builtin_offsetof(exprt &expr)
...
       typecheck_expr(index);
 
-      exprt sub_size=size_of_expr(type.subtype(), *this);
+      auto sub_size_opt = size_of_expr(type.subtype(), *this);
+
+      if(!sub_size_opt.has_value())
+      {
+        error().source_location = expr.source_location();
+        error() << "offsetof failed to determine array element size" << eom;
+        throw 0;
```
Comment (hannes-steffenhagen-diffblue) [inline src/ansi-c/c_typecheck_expr.cpp:704]: Do not throw unstructured exceptions

File: src/ansi-c/c_typecheck_expr.cpp [hunk 2]
```diff
@@ -966,18 +973,20 @@ void c_typecheck_baset::typecheck_expr_sizeof(exprt &expr)
   {
     // This is a gcc extension.
     // https://gcc.gnu.org/onlinedocs/gcc-4.8.0/gcc/Pointer-Arith.html
-    new_expr = size_of_expr(char_type(), *this);
+    new_expr = from_integer(1, size_type());
```
Comment (hannes-steffenhagen-diffblue) [inline src/ansi-c/c_typecheck_expr.cpp:976]: This appears to be a correct, but unrelated change.

File: src/ansi-c/c_typecheck_expr.cpp [hunk 3]
```diff
@@ -966,18 +973,20 @@ void c_typecheck_baset::typecheck_expr_sizeof(exprt &expr)
...
   {
-    new_expr = size_of_expr(type, *this);
+    auto size_of_opt = size_of_expr(type, *this);
 
-    if(new_expr.is_nil())
+    if(!size_of_opt.has_value())
     {
       error().source_location = expr.source_location();
       error() << "type has no size: " << to_string(type) << eom;
       throw 0;
```
Comment (hannes-steffenhagen-diffblue) [inline src/ansi-c/c_typecheck_expr.cpp:986]: This looks like it should be an invariant, judging from the error message.

File: src/ansi-c/c_typecheck_type.cpp [hunk 4]
```diff
@@ -693,11 +693,19 @@ void c_typecheck_baset::typecheck_vector_type(typet &type)
...
-  exprt sub_size_expr = size_of_expr(subtype, *this);
+  auto sub_size_expr_opt = size_of_expr(subtype, *this);
 
-  simplify(sub_size_expr, *this);
+  if(!sub_size_expr_opt.has_value())
+  {
+    error().source_location = source_location;
+    error() << "failed to determine size of vector base type `"
+            << to_string(subtype) << "'" << eom;
+    throw 0;
```
Comment (hannes-steffenhagen-diffblue) [inline src/ansi-c/c_typecheck_type.cpp:703]: Don’t throw unstructured  exceptions

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
It seems like an improvement in the sense that it is now more obvious that these functions might not return a value; but I’ve not really seen a case here where the caller can do anything sensible with the information other than ssserting that it succeeded. Maybe these functions should just `PRECONDITION` assert they’re given a sensible input (what kind of things do not have sizes?), and `POSTCONDITION` assert that their result is not nil instead?

If not, can you give an example where these functions sometimes not returning valid sizes is expected behaviour?

File: src/ansi-c/c_typecheck_expr.cpp (hunk 1 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:704]: Fair point, but the C and C++ front-ends do this all over the place at the moment. It should be fixed, but it's outside the scope of this PR.

File: src/ansi-c/c_typecheck_expr.cpp (hunk 2 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:976]: It's getting rid of one use of `size_of_expr`, which would now entail the use of `optionalt`

File: src/ansi-c/c_typecheck_expr.cpp (hunk 3 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:986]: It is, but it's something that can be triggered by user input:
````c
struct S;
int main()
{
  return sizeof(S);
}
````
which also is the reason why we cannot make `size_of_expr` fail invariants.

**Review by tautschnig — APPROVED** [review]
I believe @hannes-steffenhagen-diffblue highlighted some weaknesses, but those should be dealt with in another PR IMHO.

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
Thanks @tautschnig for addressing my comments

File: jbmc/src/java_bytecode/remove_java_new.cpp [hunk 5]
```diff
@@ -86,12 +86,12 @@ goto_programt::targett remove_java_newt::lower_java_new(
...
   // build size expression
-  exprt object_size = size_of_expr(object_type, ns);
-  CHECK_RETURN(object_size.is_not_nil());
+  const auto object_size = size_of_expr(object_type, ns);
+  CHECK_RETURN(object_size.has_value());
 
   // we produce a malloc side-effect, which stays
   side_effect_exprt malloc_expr(ID_allocate, rhs.type(), location);
-  malloc_expr.copy_to_operands(object_size);
+  malloc_expr.copy_to_operands(object_size.value());
```
Comment (smowton) [inline jbmc/src/java_bytecode/remove_java_new.cpp]: No need to change these, but for future reference you can use `operator*` for brevity

File: src/solvers/lowering/byte_operators.cpp [hunk 6]
```diff
@@ -711,16 +711,18 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
...
-        upper_bound,
-        typecast_exprt::conditional_cast(src.offset(), upper_bound.type())),
+        upper_bound_opt.value(),
+        typecast_exprt::conditional_cast(
+          src.offset(), upper_bound_opt.value().type())),
       ns);
 
   const auto lower_bound_or_nullopt = numeric_cast<mp_integer>(src.offset());
-  const auto upper_bound_or_nullopt = numeric_cast<mp_integer>(upper_bound);
+  const auto upper_bound_or_nullopt =
```
Comment (smowton) [inline src/solvers/lowering/byte_operators.cpp]: Other things like this are just called `_opt` rather than `_or_nullopt`

**Review by smowton — APPROVED** [review]

File: src/solvers/lowering/byte_operators.cpp (hunk 6 above)
Comment (kroening) [reply src/solvers/lowering/byte_operators.cpp]: changed

File: jbmc/src/java_bytecode/remove_java_new.cpp (hunk 5 above)
Comment (kroening) [reply jbmc/src/java_bytecode/remove_java_new.cpp]: sure

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 93e1a704b72e4e457b8e0aced6103c64bfaf833f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102480203
